### PR TITLE
fix(sns): create subscription object under subscriber

### LINKED ISSUE #1534: topic.subscribeQueue across stacks creates a cyclic dependency
Repro:

```ts
import cdk = require('@aws-cdk/cdk');
import sns = require('@aws-cdk/aws-sns');
import sqs = require('@aws-cdk/aws-sqs');

const app = new cdk.App();

const stackWithTopic = new cdk.Stack(app, 'stack-with-topic');
const stackWithQueue = new cdk.Stack(app, 'stack-with-queue');

const queue = new sqs.Queue(stackWithQueue, 'MyQueue');
const topic = new sns.Topic(stackWithTopic, 'MyTopic');

topic.subscribeQueue(queue);

app.run();
```

Error:

```
Stack 'stack-with-topic' already depends on stack 'stack-with-queue'. Adding this dependency would create a cyclic reference.
```

This is because `sns.Subscription` references the queue and `sqs.QueuePolicy` references the topic. I wonder if the right solution to this is to create the subscription in the scope of the queue and not in the scope of the topic.

### LINKED ISSUE #1643: Unexpected stack circular dependency when subscribing a lambda in one stack to a SNS topic in another stack
The following code is attempting to make an SNS topic in one stack, and a Lambda in another stack 

```ts
const cdk = require('@aws-cdk/cdk');
const sns = require('@aws-cdk/aws-sns');
const lambda = require('@aws-cdk/aws-lambda');
const lambdaEvents = require('@aws-cdk/aws-lambda-event-sources');

class SharedResources extends cdk.Stack {
  constructor(parent, id, props) {
    super(parent, id, props);

    this.toCrawlTopic = new sns.Topic(this, 'test-topic', {
      displayName: 'Topic'
    });
  }
}

class SubscribedLambda extends cdk.Stack {
  constructor(parent, id, props) {
    super(parent, id, props);

    const subscribedLambda = new lambda.Function(this, 'crawl-lambda', {
      runtime: lambda.Runtime.NodeJS810,
      handler: 'crawl.handle',
      code: lambda.Code.asset('./app'),
    });

    subscribedLambda.addEventSource(new lambdaEvents.SnsEventSource(props.toCrawlTopic));
  }
}

class App extends cdk.App {
  constructor(argv) {
    super(argv);

    const sharedResources = new SharedResources(this, 'shared-resources');

    new SubscribedLambda(this, 'npm-follower', {
      toCrawlTopic: sharedResources.toCrawlTopic
    });
  }
}

new App().run();
```

However when I deploy I get this error:

```
Error: Stack 'shared-resources' already depends on stack 'npm-follower'. Adding this dependency would create a cyclic reference.
    at SharedResources.addDependency (/Users/peckn/Documents/Code/test-case/node_modules/@aws-cdk/cdk/lib/cloudformation/stack.js:185:19)
    at CfnReference.consumeFromStack (/Users/peckn/Documents/Code/test-case/node_modules/@aws-cdk/cdk/lib/cloudformation/cfn-tokens.js:53:28)
    at SharedResources.prepare (/Users/peckn/Documents/Code/test-case/node_modules/@aws-cdk/cdk/lib/cloudformation/stack.js:334:21)
    at ConstructNode.prepareTree (/Users/peckn/Documents/Code/test-case/node_modules/@aws-cdk/cdk/lib/core/construct.js:248:27)
    at App.synthesizeStack (/Users/peckn/Documents/Code/test-case/node_modules/@aws-cdk/cdk/lib/app.js:55:19)
    at App.synthesizeStacks (/Users/peckn/Documents/Code/test-case/node_modules/@aws-cdk/cdk/lib/app.js:85:27)
    at App.run (/Users/peckn/Documents/Code/test-case/node_modules/@aws-cdk/cdk/lib/app.js:43:26)
    at Object.<anonymous> (/Users/peckn/Documents/Code/test-case/minimal.js:42:68)
    at Module._compile (module.js:649:30)
    at Object.Module._extensions..js (module.js:660:10)
```

My expectation is that the `npm-follower` stack would get a dependency on the `shared-resources` stack but not the other way around. Somehow it becomes a circular dependency though

### PR DESCRIPTION
**Pull request by rix0rrr**
Create the SNS Subscription object under the subscriber
instead of the subscribee, avoiding cyclic stack dependencies
if those objects would be in different stacks.

Fixes #1643, fixes #1534.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-sns/lib/topic-ref.ts [hunk 1]
```diff
@@ -169,14 +169,18 @@ export abstract class TopicBase extends cdk.Construct implements ITopic {
    * @param queue The target queue
    */
   public subscribeQueue(queue: sqs.IQueue): Subscription {
-    const subscriptionName = queue.node.id + 'Subscription';
-    if (this.node.tryFindChild(subscriptionName)) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-sns/lib/topic-ref.ts]: We should make IQueue extend IConstruct

File: packages/@aws-cdk/aws-sns/lib/topic-ref.ts [hunk 2]
```diff
@@ -169,14 +169,18 @@ export abstract class TopicBase extends cdk.Construct implements ITopic {
...
+
+    const subscriptionName = this.node.id + 'Subscription';
+    if (queue.node.tryFindChild(subscriptionName)) {
       throw new Error(`A subscription between the topic ${this.node.id} and the queue ${queue.node.id} already exists`);
     }
 
     // we use the queue name as the subscription's. there's no meaning to subscribing
     // the same queue twice on the same topic.
-    const sub = new Subscription(this, subscriptionName, {
+    const sub = new Subscription(queue, subscriptionName, {
```
Comment (eladb) [inline packages/@aws-cdk/aws-sns/lib/topic-ref.ts]: Explain why you use the queue as a scope

**Comment by eladb** [comment]
Make sure to indicate this is a breaking change since it will cause resource replacement

File: packages/@aws-cdk/aws-sns/lib/topic-ref.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-sns/lib/topic-ref.ts]: It does, but `scope: Construct` and not `scope: IConstruct`.

Comment (eladb) [reply packages/@aws-cdk/aws-sns/lib/topic-ref.ts]: Damn, right. I feel this is going to bite us at some point. But I really don't want people to need to specify `scope: IConstruct` in the initializer...

**Comment by eladb** [comment]
> We also need a different term for "impacts existing deployments" than BREAKING CHANGE I feel.

You are right. It's going to recreate a resource, but without data or availability loss, so I am okay with not indicating as a breaking change. Maybe we can devise some sort of a code in the commit title that we can later on expand in CHANGELOG?

**Comment by rix0rrr** [comment]
>  Maybe we can devise some sort of a code in the commit title that we can later on expand in CHANGELOG?

That's what I was thinking as well. What would be a good term though? DEPLOYMENT IMPACT?

**Pull request merged**